rfcs/20211025-augru/README.md

```diff
@@ -0,0 +1,1089 @@
+Proposal for GRU with Attentional Update in DNNL
+================================================
+
+## 1. Introduction
+
+In addition to the regular GRU there is a notion of GRU with attentional update
+gate (AUGRU) that can help to model interest evolving in Click-Through Rate
+Prediction [[#1]][1].
+
+
+### 1.1. GRU
+
+GRU is defined by the following formula:
+
+![GRU-LBR](formula-gru-lbr.png)
+
+
+### 1.2. AUGRU
+
+AUGRU is defined by the following formula (the difference from the
+regular GRU is highlighted in red):
+
+![AUGRU-LBR](formula-augru-lbr.png)
+
+
+
+> Warning
+>
+> Attention is passed not as part of source layer as it is done in Neural Machine
+> Translation (NMT) models [[#3]][3], but as a separate memory which is
+> applied to intermediate results (to gate U). This difference makes it impossible
+> to apply attention on application side and pass the result to oneDNN RNN
+> primitive.
+
+## 2. API Changes
+
+Attention memory descriptor can be built from existing `rnn_desc_t` fields (memory
+descriptors, cell kind, direction, propagation kind, etc.). Taking into account
+this fact and to avoid breaking changes an attention memory descriptor will not
+be part of `dnnl_rnn_desc_t`. Instead a user should use query API to get
+attention shapes.
+
+ * Pros:
+   - RNN primitive descriptor can request the best layout for an attention memory
+   - RNN primitive descriptor can request the data type compatible with RNN cell
+ * Cons:
+   - User has to reorder/quantize their attention data to align with the one
+   requested by RNN
+ * Alternative:
+   - Introduce `rnn_desc_v2_t` with a memory descriptor for an attention.
```

Comment:
It sounds like this alternative might be more user-friendly. Would the drawback be the extra complexity of defining the new rnn descriptor and memory descriptor? Sort of like how there are two versions of pooling?

Reply:
Correct, the downside would be supporting both rnn_desc and rnn_desc_v2 for backward compatibility. Actually it is not clear why user might want to have that descriptor, because sizes of attention are tied to other memory descriptors and its data type is tied to dst data type. The only reason would be memory format, which anyway we would ask to pass as any so oneDNN will return the most convenient one for the implementations which is similar to the current proposal. Another issue is that attention can mean different things for different RNNs, so we will have to add algorithm kind along with memory descriptor. As a result I am not sure that adding memory descriptor will bring any benefit to users, but will bring complexity to the library.

I plan to share an implementation of AUGRU with framework teams so they can try it and provide a feedback before we promote the changes.

Reply:
Makes sense, thanks for the explanation! Pending feedback from the frameworks team, it does sound like the complexity would be much greater than the benefit to the user.